### PR TITLE
Fix optimization, transformed data prints

### LINKED ISSUE #269: Prints in the transform data are not printed
As reported in https://github.com/stan-dev/stanc3/issues/652, 
print() in transform data does not get printed.

The reason for this is that in order to not print the initial Cmdstan metadata dump for all chains, we start printing stdout to the user only after either an error occurs, an init warning is read or the start of warmup/sampling is printed. So not ideal.

For the time being, we might get away with printing everything after the line
```
refresh = x
```
which I think is guaranteed to always be the last metadata printed (do have to check though). But that means that we will always print the gradient times for all chains. Hopefully, we can parse that away.

The ideal solution for this would be the "wrapper-friendly" mode for Cmdstan. But that won't be available at least until next release if we get others on board with that idea.

### LINKED ISSUE #314: The optimize() function doesn't return output when it fails to converge
**Describe the bug**

When running `optimize` using CmdStan, it can return a value even if a global optimum isn't reached.

1. CmdStan does write to file, even if algorithm diagnostics state that it may not have found the right optimum.
2. RStan currently returns the value.
3. CmdStanR does not. It does report the right information.

**To Reproduce**

I used 8 schools and ran optimization on it.

From R (after setting up `model`):
```
> fit = model$optimize(data = schools_dat)
method = optimize 
  optimize 
    algorithm = lbfgs (Default) 
      lbfgs 
        init_alpha = 0.001 (Default) 
        tol_obj = 9.9999999999999998e-13 (Default) 
        tol_rel_obj = 10000 (Default) 
        tol_grad = 1e-08 (Default) 
        tol_rel_grad = 10000000 (Default) 
        tol_param = 1e-08 (Default) 
        history_size = 5 (Default) 
    iter = 2000 (Default) 
    save_iterations = 0 (Default) 
id = 1 
data 
  file = /tmp/RtmplVvoS1/standata-14844a9babf.json 
init = 2 (Default) 
random 
  seed = 1389664808 
output 
  file = /tmp/RtmplVvoS1/eight_schools_4-202010090300-1-a51a43.csv 
  diagnostic_file =  (Default) 
  refresh = 100 (Default) 
 
Initial log joint probability = -47.8461 
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes  
      99       228.766        0.6228   8.01767e+13      0.3575      0.3575      259    
    Iter      log prob        ||dx||      ||grad||       alpha      alpha0  # evals  Notes  
     118       252.377   0.000284799   9.72824e+14       1e-12       0.001      387  LS failed, Hessian reset  
Optimization terminated with error:  
  Line search failed to achieve a sufficient decrease, no more progress can be made 
Warning: Fitting finished unexpectedly!

> fit
Error: Fitting failed. Unable to print.
> 
```

**Expected behavior**

Even though it isn't able to find an optimum, it should return what's in `output.csv`.

Compare this to rstan:
```
> fit = rstan::stan("eight_schools.stan", data = schools_dat)
...
> optimum = rstan::optimizing(fit@stanmodel, data = schools_dat)
Warning message:
In .local(object, ...) : non-zero return code in optimizing
> optimum
$par
    theta[1]     theta[2]     theta[3]     theta[4]     theta[5]     theta[6]     theta[7]     theta[8]           mu 
1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 1.288251e+01 
         tau 
3.353735e-15 

$value
[1] 240.5844

$return_code
[1] 70

$theta_tilde
     theta[1] theta[2] theta[3] theta[4] theta[5] theta[6] theta[7] theta[8]       mu          tau
[1,] 12.88251 12.88251 12.88251 12.88251 12.88251 12.88251 12.88251 12.88251 12.88251 3.353735e-15

> 
```

**Operating system**
RStudio Cloud

**CmdStanR version number**
```
> packageVersion("cmdstanr")
[1] ‘0.1.3’
```

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Summary

Fixes #314
Fixes #269

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar, Uni. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/run.R [hunk 1]
```diff
@@ -390,7 +390,15 @@ CmdStanRun$set("private", name = "run_generate_quantities_", value = .run_genera
     }
     procs$set_active_procs(procs$num_alive())
   }
-  if (procs$get_proc(id)$get_exit_status() == 0) {
+  successful_fit = FALSE
+  if (self$method() == "optimize") {
+    if (procs$proc_state(id = id) > 3) {
```
Comment (rok-cesnovar) [inline R/run.R:395]: This marks that either Optimization terminated with error or Optimization terminated normally was printed which means that optimization terminated "normally".

File: R/run.R [hunk 2]
```diff
@@ -663,7 +691,11 @@ CmdStanMCMCProcs <- R6::R6Class(
           # (Note: state 2 is only used because rejection in cmdstan is printed
           # to stdout not stderr and we want to avoid printing the intial chain metadata)
           next_state <- state
-          if (state < 3 && regexpr("Rejecting initial value:", line, perl = TRUE) > 0) {
+          if (state < 3 && regexpr("refresh =", line, perl = TRUE) > 0) {
```
Comment (rok-cesnovar) [inline R/run.R:694]: Once we find `refresh = 0`, we wait for a blank line and that marks the end of "metadata". This fixes the "sig_figs = -1" prints bug.

File: R/run.R [hunk 3]
```diff
@@ -684,15 +716,25 @@ CmdStanMCMCProcs <- R6::R6Class(
...
           if (regexpr("seconds (Warm-up)", line, fixed = TRUE) > 0) {
             private$proc_section_time_[id, "warmup"] <- as.double(trimws(sub("Elapsed Time: ", "", sub("seconds (Warm-up)", "", line, fixed = TRUE), fixed = TRUE)))
+            next_state <- 5
+            state <- 5
           }
-          if (state > 1 && state < 5) {
+          if (regexpr("Gradient evaluation took",line, fixed = TRUE) > 0
+              || regexpr("leapfrog steps per transition would take",line, fixed = TRUE) > 0
+              || regexpr("Adjust your expectations accordingly!",line, fixed = TRUE) > 0) {
+            ignore_line <- TRUE
```
Comment (rok-cesnovar) [inline R/run.R:735]: We now print everything after metadata. I decided to mask the above lines that I am not sure are very helpful messages and mostly just spam the cmdline.

File: tests/testthat/test-fit-shared.R [hunk 4]
```diff
@@ -251,8 +251,8 @@ test_that("CmdStanArgs erorrs if idx is out of proc_ids range", {
 
 test_that("no output with refresh = 0", {
   skip_on_cran()
-  mod <- testing_model("bernoulli")
-  data_list <- testing_data("bernoulli")
+  mod <- testing_model("logistic")
+  data_list <- testing_data("logistic")
```
Comment (rok-cesnovar) [inline tests/testthat/test-fit-shared.R:255]: The bernoulli variational example was failing every once in a while.

**Review by jgabry — APPROVED** [review]
Awesome, thanks. I just added two tests, one checking that prints in transformed data work (wasn't sure where to put that test) and one checking that estimates are accessible even after optimization fails. I won't merge yet in case you any problems with the tests I added, but if they look good go ahead and merge!

**Comment by rok-cesnovar** [comment]
Thanks for adding tests. Looks good to me!

**Comment by jgabry** [comment]
Ok cool. I guess let's let the checks finish running and then merge if everything passes.

**Pull request merged**